tests/base/src/test/java/org/wildfly/security/ssl/SSLAuthenticationTest.java

```diff
@@ -731,60 +725,93 @@ public void testClientSideOcsp() throws Exception {
 
         AuthenticationContext context = AuthenticationContext.getContextManager().get();
         AuthenticationContextConfigurationClient contextConfigurationClient = AccessController.doPrivileged(AuthenticationContextConfigurationClient.ACTION);
         SSLContext clientContext = contextConfigurationClient.getSSLContext(URI.create(clientUri), context);
 
-        return performConnectionTest(serverContext, clientContext, expectValid);
+        try {
+            testCommunication(serverContext, clientContext, expectedServerPrincipal, expectedClientPrincipal, expectedCipherSuite);
+            if(!expectValid) Assert.fail("Expected SSLHandshakeException not thrown");
+        } catch (SSLHandshakeException expected) {
+            //expected
+        } catch (SSLException e) {
+            if (e.getCause() instanceof SocketException) return; //expected
+            throw new IllegalStateException("Unexpected SSLException", e);
+        }
     }
 
-    private SecurityIdentity performConnectionTest(SSLContext serverContext, SSLContext clientContext, boolean expectValid) throws Exception {
-        SSLServerSocketFactory sslServerSocketFactory = serverContext.getServerSocketFactory();
-
-        SSLServerSocket sslServerSocket = (SSLServerSocket) sslServerSocketFactory.createServerSocket(1111, 10, InetAddress.getLoopbackAddress());
+    private void testCommunication(SSLContext serverContext, SSLContext clientContext, String expectedServerPrincipal, String expectedClientPrincipal, String expectedCipherSuite) throws Throwable {
+        ServerSocket listeningSocket = serverContext.getServerSocketFactory().createServerSocket();
+        listeningSocket.bind(new InetSocketAddress("localhost", TESTING_PORT));
+        SSLSocket clientSocket = (SSLSocket) clientContext.getSocketFactory().createSocket("localhost", TESTING_PORT);
+        SSLSocket serverSocket = (SSLSocket) listeningSocket.accept();
 
-        ExecutorService executorService = Executors.newSingleThreadExecutor();
-        Future<SSLSocket> socketFuture = executorService.submit(() -> {
+        ExecutorService serverExecutorService = Executors.newSingleThreadExecutor();
+        Future<byte[]> serverFuture = serverExecutorService.submit(() -> {
             try {
-                System.out.println("About to connect client");
-                SSLSocket sslSocket = (SSLSocket) clientContext.getSocketFactory().createSocket(InetAddress.getLoopbackAddress(), 1111);
-                sslSocket.getSession();
+                byte[] received = new byte[2];
+                serverSocket.getInputStream().read(received);
+                serverSocket.getOutputStream().write(new byte[]{0x56, 0x78});
 
-                return sslSocket;
+                if (expectedClientPrincipal != null) {
+                    Assert.assertEquals(expectedClientPrincipal, serverSocket.getSession().getPeerPrincipal().getName());
+                }
+
```

Comment:
You should be able to use ```(SecurityIdentity) serverSocket.getSession().getValue(SSLUtils.SSL_SESSION_IDENTITY_KEY)``` to do the ```assertNull``` / ```assertNotNull``` checks on the identity here.

Reply:
Added, thanks!